homie-controller/src/lib.rs

```diff
@@ -0,0 +1,407 @@
+        device_id: String,
+        node_id: String,
+        has_required_attributes: bool,
+    },
+    PropertyUpdated {
+        device_id: String,
+        node_id: String,
+        property_id: String,
+        has_required_attributes: bool,
+    },
+    PropertyValueChanged {
+        device_id: String,
+        node_id: String,
+        property_id: String,
+        value: String,
+    },
+}
+
+impl Event {
+    fn device_updated(device: &Device) -> Self {
+        Event::DeviceUpdated {
+            device_id: device.id.to_owned(),
+            has_required_attributes: device.has_required_attributes(),
+        }
+    }
+
+    fn node_updated(device_id: &str, node: &Node) -> Self {
+        Event::NodeUpdated {
+            device_id: device_id.to_owned(),
+            node_id: node.id.to_owned(),
+            has_required_attributes: node.has_required_attributes(),
+        }
+    }
+
+    fn property_updated(device_id: &str, node_id: &str, property: &Property) -> Self {
+        Event::PropertyUpdated {
+            device_id: device_id.to_owned(),
+            node_id: node_id.to_owned(),
+            property_id: property.id.to_owned(),
+            has_required_attributes: property.has_required_attributes(),
+        }
+    }
+
+    fn property_value(device_id: &str, node_id: &str, property: &Property) -> Self {
+        Event::PropertyValueChanged {
+            device_id: device_id.to_owned(),
+            node_id: node_id.to_owned(),
+            property_id: property.id.to_owned(),
+            value: property.value.to_owned().unwrap(),
+        }
+    }
+}
+
+/// A Homie controller, which connects to an MQTT broker and interacts with Homie devices.
+#[derive(Debug)]
+pub struct HomieController {
+    mqtt_client: AsyncClient,
+    base_topic: String,
+    /// The set of Homie devices which have been discovered so far, keyed by their IDs.
+    pub devices: Arc<Mutex<HashMap<String, Device>>>,
+}
+
+impl HomieController {
+    /// Create a new `HomieController` connected to an MQTT broker.
+    ///
+    /// # Arguments
+    /// * `base_topic`: The Homie [base topic](https://homieiot.github.io/specification/#base-topic)
+    ///   under which to look for Homie devices. "homie" is the recommended default.
+    /// * `mqtt_options`: Options for the MQTT connection, including which broker to connect to.
+    pub fn new(mqtt_options: MqttOptions, base_topic: &str) -> (HomieController, EventLoop) {
+        let (mqtt_client, event_loop) = AsyncClient::new(mqtt_options, REQUESTS_CAP);
+        let controller = HomieController {
+            mqtt_client,
+            base_topic: base_topic.to_string(),
+            devices: Arc::new(Mutex::new(HashMap::new())),
+        };
+        (controller, event_loop)
+    }
+
+    /// Poll the `EventLoop`, and maybe return a Homie event.
+    pub async fn poll(&self, event_loop: &mut EventLoop) -> Result<Option<Event>, PollError> {
+        let notification = event_loop.poll().await?;
```

Comment:
I think the problem is that we're handing out event_loop in a way that our api consumer can decide whether they poll it directly or poll it via this method.

Could we wrap it in an opaque type that has a poll method?

It could be the simple reverse of this function signature (`HomieEventLoop::poll(&mut self, &HomieController)`) or HomieEventLoop could hold a reference to HomieController somehow, and make it `HomieEventLoop::poll(&mut self)`

Reply:
I can't figure out how to make `HomieEventLoop` include a reference to `HomieController`, unless we always wrap `HomieController` in an `Arc`. Maybe we can talk this through some more.

Reply:
I can't see a way around Arc either. Maybe it's something to think about in another pr.

Reply:
Wrapped in an opaque type for now. I'll consider an Arc on `HomieEventLoop` or a Mutex on `HomieController` later.